src/qps/interface/qps.c

```diff
@@ -683,14 +681,19 @@ PetscErrorCode QPSGetConvergenceContext(QPS qps,void **ctx)
 .seealso: QPSSetConvergenceTest(), QPSSetTolerances(), QPSConvergedSkip(), KSPConvergedReason, QPSGetConvergedReason(),
           QPSConvergedDefaultCreate(), QPSConvergedDefaultDestroy()
 @*/
-PetscErrorCode QPSConvergedDefault(QPS qps,QP qp,PetscInt i,PetscReal rnorm,KSPConvergedReason *reason,void *ctx)
+PetscErrorCode QPSConvergedDefault(QPS qps,KSPConvergedReason *reason,void *ctx)
 {
   QPSConvergedDefaultCtx *cctx = (QPSConvergedDefaultCtx*) ctx;
+  QP qp = qps->solQP;
+  PetscInt i = qps->iteration;
+  PetscReal rnorm = qps->rnorm;
 
   PetscFunctionBegin;
   PetscValidHeaderSpecific(qps,QPS_CLASSID,1);
+  PetscValidHeaderSpecific(qp,QP_CLASSID,2);
   *reason = KSP_CONVERGED_ITERATING;
 
+  if (!qps->setupcalled) FLLOP_SETERRQ(((PetscObject) qps)->comm, PETSC_ERR_ARG_WRONGSTATE, "QPSSetUp() not yet called");
```

Comment:
Again, is this necessary?

Reply:
Here it's even more important. We need to be sure e.g. that `solQP` is in place.

Reply:
OK. However, I would suggest putting this in QPSConvergedDefaultSetUp(), which also computes norm of solQP->b and then this check will be done only once.

Reply:
OK, this makes sense.